### PR TITLE
Make join_segmentations return array maps from output to input labels

### PR DESCRIPTION
**Pull request by aeisenbarth**
## Description

This is an implementation for https://github.com/scikit-image/scikit-image/issues/6785.

It adds as optional return value two ArrayMaps which allow tracing the labels of the returned segmentation back to the labels of the two input segmentations.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

File: skimage/segmentation/_join.py
Comment (aeisenbarth) [inline skimage/segmentation/_join.py:60]: The original code was reassigning modified data to the same references.
Here, `labels_joined = np.unique(j)` must be computed before relabeöing `j`.

- Usually I would use unique reference names to avoid depending on the correct order of lines (and potential breakages). I don't know whether `j`was overwritten for performance/memory reasons.
- Also, the unique operation is only needed when return_mapping is True. I don't know what performance impact it has when running the function with its defaults. One could move `labels_joined` into a conditional, but then the reference is potentially undefined, unless we also define it in an else branch (`else: labels_joined = None`)

**Comment by aeisenbarth** [comment]
- In my original code, I used more expressive names also for `s1`, `j` etc. I can changed variable names if desired.
- I would also add type hints (although the existing code didn't have).

File: skimage/segmentation/_join.py
Comment (jni) [inline skimage/segmentation/_join.py:61]: I think it's impossible for this to fail? So might as well remove it?

File: skimage/segmentation/_join.py
Comment (jni) [reply skimage/segmentation/_join.py:56]: > Usually I would use unique reference names to avoid depending on the correct order of lines (and potential breakages). I don't know whether `j` was overwritten for performance/memory reasons.

Nope! It was rewritten willy-nilly. 😅 I wrote this code ages ago when I was much less experienced. I 100% agree about not reusing reference names and support changing the first reference to e.g. `j_initial`? I'm also ok with using more descriptive names throughout.

> Also, the unique operation is only needed when return_mapping is True. I don't know what performance impact it has when running the function with its defaults.

But if you don't overwrite `j`, you can move it to after the return clause and all is well, right?

File: skimage/segmentation/_join.py
Comment (jni) [inline skimage/segmentation/_join.py:45]: This is hard to read for someone who doesn't already know about `ArrayMap`, so I might instead suggest:

```python
>>> j, m1, m2 = join_segmentations(s1, s2, return_mapping=True)
>>> m1
ArrayMap(array([0, 1, 2, 3, 4, 5]), array([0, 0, 1, 1, 2, 2]))
>>> np.all(m1[j] == s1)
True
>>> np.all(m2[j] == s2)
True
```

What do you think?

**Review by jni — CHANGES_REQUESTED** [review]
@aeisenbarth very cool!

In the long term for scikit-image, we are moving away from `return_something=` arguments that change the return type, because they make typing very awkward. However, since we haven't quite decided yet what we'll replace it with 😅, I'm personally in favour of merging this. It's a very nice addition. I have left a couple of suggestions, but otherwise I'll approve once fixed. Thanks!!!

File: skimage/segmentation/_join.py
Comment (aeisenbarth) [reply skimage/segmentation/_join.py]: > But if you don't overwrite j, you can move it to after the return clause and all is well, right?
Yes, I renamed the variables (hopefully not too verbose) and moved the line after the return.

Comment (aeisenbarth) [reply skimage/segmentation/_join.py]: > What do you think?

That example is much better, I updated it.

File: skimage/segmentation/_join.py
Comment (aeisenbarth) [reply skimage/segmentation/_join.py]: Yes, I renamed the variables (hopefully not too verbose) and moved the line after the return.

**Comment by aeisenbarth** [comment]
Shall I also add type hints? (I don't know whether scikit-image has any constraints on compatibility with older Python syntax).
Does scikit-image have any preferred array type hints?

Something like this should correctly describe the optional return values:
```python
@overload
def join_segmentations(
    s1: np.ndarray, s2: np.ndarray, return_mapping: Literal[False] = False
) -> np.ndarray:
    ...


@overload
def join_segmentations(
    s1: np.ndarray, s2: np.ndarray, return_mapping: Literal[True]
) -> Tuple[np.ndarray, ArrayMap, ArrayMap]:
    ...


def join_segmentations(
    s1: np.ndarray, s2: np.ndarray, return_mapping: bool = False
) -> Union[np.ndarray, Tuple[np.ndarray, ArrayMap, ArrayMap]]:
```

**Comment by lagru** [comment]
Although, long-term type hints are the direction we want to go, right now they wouldn't be consistent with the rest of the library. So I'd say we don't need to add them in this PR. 

Thanks for introducing me to the [`overload` decorator](https://docs.python.org/3/library/typing.html#typing.overload)! :blush: That could be quite useful down the line.

File: skimage/segmentation/_join.py
Comment (lagru) [inline skimage/segmentation/_join.py:6]: What do you guys think about having two functions to avoid the "may return different output lengths"-mess? E.g.

```python
def join_segmentation_map(s1, s2):
    ...
    return j, map_j_to_s1, map_j_to_s2

def join_segmentations(s1, s2):
    j, *_ = join_segmentation_map(s1, s2)
    return j
```

Regardless of how we decide to tackle this issue in the future, I think this approach is easier to update with deprecations and doesn't introduce a lot of duplication, besides the docstrings.

Comment (jni) [reply skimage/segmentation/_join.py:6]: Nope, hate it LOL. 

Since the default is False I don't think deprecations text backwards compatibility are an issue. Going forward, I think either approach is equally likely to result in a deprecation. And the two functions are (imho) ugly and I don't see us keeping it long term. One close alternative would be join_segmentations vs extra.join_segmentations (ie new namespace, same function name), but then you've still got a deprecation anyway.

Either way, I think this is a much bigger discussion that shouldn't hold up this PR. For now, `return_foo=` is actually the most consistent with the rest of the library, so I think we should stick with it.

**Review by jni — APPROVED** [review]

File: skimage/segmentation/_join.py
Comment (lagru) [reply skimage/segmentation/_join.py:6]: I can live with that. :smile: I find it interesting that you have such a strong reaction to it. For me that pattern seems way more expressive and easier to reason about / test. But I agree that's an interesting discussion for another time.

**Comment by lagru** [comment]
> Shall we merge this then? =)

Happy to. Just didn't get around to review this yet...

**Comment by jni** [comment]
@lagru you're a busy man right now so I decided to save you some trouble 😂

**Pull request merged**
